File: apps/yukino/tests/team-file.test.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { existsSync, mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { describe, expect, test, beforeEach, afterEach } from "vitest";

import { TeamManager } from "@/teams/index.js";
import { readTeamFile, teamDir, teamsBaseDir } from "@/teams/team-file.js";

// The teams directory lives at <home>/.yukino/teams, so redirect the entire
// home directory to a temp dir to avoid leaving residue in the real
// ~/.yukino/teams. os.homedir() reads USERPROFILE on Windows and HOME on other
// platforms, so set both.
let origHome: string | undefined;
let origUserProfile: string | undefined;

beforeEach(() => {
  origHome = process.env.HOME;
  origUserProfile = process.env.USERPROFILE;
  const tmp = mkdtempSync(join(tmpdir(), "yukino-home-"));
  process.env.HOME = tmp;
  process.env.USERPROFILE = tmp;
});

afterEach(() => {
  if (origHome === undefined) {
    delete process.env.HOME;
  } else {
    process.env.HOME = origHome;
  }
  if (origUserProfile === undefined) {
    delete process.env.USERPROFILE;
  } else {
    process.env.USERPROFILE = origUserProfile;
  }
});

const workDir = () => mkdtempSync(join(tmpdir(), "yukino-work-"));

describe("team config persistence", () => {
  test("can be read back by a fresh TeamManager after writing to disk", () => {
    const mgr = new TeamManager(workDir());
    const team = mgr.create("Refactor Auth", "in-process", {
      leadAgentId: "lead",
      description: "Refactor the authentication module",
    });
    team.addMember("alice");
    team.setMemberMeta("alice", {
      agentType: "worker",
      model: "claude-sonnet-4-6",
      worktreePath: "/tmp/wt/alice",
    });

    // Swap in a brand-new manager to simulate a teammate process or the next session
    const fresh = new TeamManager(workDir());
    const got = fresh.get("Refactor Auth");

    expect(got).toBeDefined();
    expect(got?.leadAgentId).toBe("lead");
    expect(got?.description).toBe("Refactor the authentication module");

    const m = got?.getMember("alice");
    expect(m).toBeDefined();
    expect(m?.agentType).toBe("worker");
    expect(m?.model).toBe("claude-sonnet-4-6");
    expect(m?.worktreePath).toBe("/tmp/wt/alice");
  });

  test("slugifies the team directory name", () => {
    const mgr = new TeamManager(workDir());
    mgr.create("Refactor Auth!", "tmux", { leadAgentId: "lead" });

    const expected = join(teamsBaseDir(), "refactor-auth-", "config.json");
    expect(existsSync(expected)).toBe(true);
  });

  test("tearing down a team removes the entire team directory", async () => {
    const mgr = new TeamManager(workDir());
    mgr.create("gone", "in-process", { leadAgentId: "lead" });
    expect(existsSync(teamDir("gone"))).toBe(true);

    await mgr.delete("gone");
    expect(existsSync(teamDir("gone"))).toBe(false);
    expect(new TeamManager(workDir()).get("gone")).toBeUndefined();
  });

  test("returns undefined for a team that does not exist", () => {
    expect(new TeamManager(workDir()).get("never-existed")).toBeUndefined();
  });

  test("persists a member's active state into the config", async () => {
    const mgr = new TeamManager(workDir());
    const team = mgr.create("t", "in-process", { leadAgentId: "lead" });
    team.addMember("bob");
    await team.stopMember("bob");

    const tf = readTeamFile("t");
    expect(tf).not.toBeNull();
    expect(tf?.members).toHaveLength(1);
    expect(tf?.members[0].isActive).toBe(false);
  });
});

File: yukino/teams/team_file_test.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package teams

import (
	"os"
	"path/filepath"
	"testing"
)

func TestTeamFileRoundTrip(t *testing.T) {
	base := t.TempDir()
	tm := NewTeamManager(base, base)
	team := tm.CreateTeamFull("Refactor Auth", "lead", "refactor the auth module")
	team.AddMember("alice")
	team.SetMemberMeta("alice", MemberMeta{
		AgentType:    "worker",
		Model:        "claude-sonnet-4-6",
		WorktreePath: "/tmp/wt/alice",
	})

	// Use a fresh manager to simulate a teammate process or the next session.
	fresh := NewTeamManager(base, base)
	got := fresh.GetTeam("Refactor Auth")
	if got == nil {
		t.Fatal("expected team to be reconstructed from disk, got nil")
	}
	if got.LeadAgentID != "lead" {
		t.Errorf("LeadAgentID = %q, want lead", got.LeadAgentID)
	}
	if got.Description != "refactor the auth module" {
		t.Errorf("Description = %q, want 'refactor the auth module'", got.Description)
	}
	m := got.GetMember("alice")
	if m == nil {
		t.Fatalf("member alice was not restored, current members: %v", got.MemberNames())
	}
	if m.AgentType != "worker" || m.Model != "claude-sonnet-4-6" || m.WorktreePath != "/tmp/wt/alice" {
		t.Errorf("member metadata mismatch: %+v", m)
	}
}

func TestTeamFilePathIsSanitized(t *testing.T) {
	base := t.TempDir()
	tm := NewTeamManager(base, base)
	tm.CreateTeamFull("Refactor Auth!", "lead", "")

	want := filepath.Join(base, "refactor-auth-", "config.json")
	if _, err := os.Stat(want); err != nil {
		t.Fatalf("expected config at %s, stat failed: %v", want, err)
	}
}

func TestDeleteTeamRemovesDir(t *testing.T) {
	base := t.TempDir()
	tm := NewTeamManager(base, base)
	tm.CreateTeamFull("gone", "lead", "")
	if _, err := os.Stat(tm.teamDir("gone")); err != nil {
		t.Fatalf("directory should exist after team creation: %v", err)
	}

	tm.DeleteTeam("gone")
	if _, err := os.Stat(tm.teamDir("gone")); !os.IsNotExist(err) {
		t.Errorf("directory should be removed after team deletion, err = %v", err)
	}
	if fresh := NewTeamManager(base, base).GetTeam("gone"); fresh != nil {
		t.Errorf("a deleted team should not be recoverable from disk")
	}
}

func TestGetTeamMissingReturnsNil(t *testing.T) {
	if got := newTmpManager(t).GetTeam("never-existed"); got != nil {
		t.Errorf("non-existent team should return nil, got %+v", got)
	}
}
